### ISSUE TITLE
[Feature request] Integration to pocketbook native system by updating percentage progress in book library

### ISSUE THREAD

**Issue by druss67** [koreader/koreader#1493]
I have noticed that koreader do not update the "percentage read" in the library of my PB626 (firm 5.7)
I tought it was due to the use of an extra reader, but it works with the fork of CoolReader. 

![scr0002](https://cloud.githubusercontent.com/assets/11350948/7216913/c850b036-e610-11e4-9211-f7ce8369bbc4.jpg)

**Comment by chrox** [comment]
It's possible to write a special program to update the progress of a document with pocketbook SDK. And call that program after we exit koreader just like what we do in kpvbooklet for kindle. But I have no pocketbook device anymore so have no motivation to implement this.

**Comment by druss67** [comment]
I understand, no problem. Thanks

**Comment by ploum** [comment]
I badly want that feature and I was wondering what was the best option for me : sponsoring @chrox by offering him a PB device or trying to get my hands dirty in the code ? Does it looks like it could be easily solved or does it requires a great experience of koreader code?

**Comment by Frenzie** [comment]
I think wrt to PB you might have more luck pinging @ezdiy sometime next month.

> Does it looks like it could be easily solved or does it requires a great experience of koreader code?

Note that there's another part to it: figuring out how the integration with the PB system works by linking to their docs on it (if there are any).

**Comment by pazos** [comment]
or ping @blchinezu, the author behind https://github.com/blchinezu/pocketbook-coolreader. According to OP:

> I tought it was due to the use of an extra reader, but it works with the fork of CoolReader.

**Comment by ploum** [comment]
Thanks for the info. I was using CoolReader before and I planned to dig a bit into their code to see how they did it. Not sure about the shape of the code but that's indeed a way I can help without digging into koreader myself.

**Comment by blchinezu** [comment]
Was I summoned?

Header file that should provide what you need: [FRSCSDK/include/bookstate.h](https://github.com/blchinezu/pocketbook-sdk/blob/master/FRSCSDK/include/bookstate.h)

CoolReader implementation: [CRPocketBookProStateSaver](https://github.com/blchinezu/pocketbook-coolreader/blob/d032f332ea2e392b04276d3c13ff07e98d4462f3/cr3gui/src/cr3pocketbook.cpp#L168)
There's also this method: [void CRPocketBookGlobals::saveState(int cpage, int npages)](https://github.com/blchinezu/pocketbook-coolreader/blob/d032f332ea2e392b04276d3c13ff07e98d4462f3/cr3gui/src/cr3pocketbook.cpp#L333)

This is quite old so I don't know if it's still working or not with the latest firmware.
Also, this part of the code in particular was not written by me but by [@pkb](https://github.com/pkb).

Hope it helps.

**Comment by ploum** [comment]
Thanks. I've dig a bit into it and it looks a bit complex my capabilites. Little chance that I could do it myself.

If anyone is seriously interested by fixing this issue, I'm ready to offer a pocketbook device.

**Comment by pazos** [comment]
> Header file that should provide what you need: FRSCSDK/include/bookstate.h
> Hope it helps.

Indeed, very useful, thanks @blchinezu

> If anyone is seriously interested by fixing this issue, I'm ready to offer a pocketbook device.

@ploum: it is still a won't fix on my end. Albeit the harder part isn't the interface with pocketbook specific functions but the integration in the application lifetime, for example:

- Shall books opened within the KO FM be updated?, all of them or just the ones supported by or associated with the app? If a book is associated with KOReader but opened on a different app, does the new app get the current page from the book handle?. If we open the book using KOReader FM shall we update the current page in spite of not knowing which will be the next book consumer? How to know if the current page in KOReader with font size 60 is <= the number of pages of the official reader with the font size 10?.

- When we need to update the book handle?. Only when we receive an EVT that returns back to the PB library? (in that case we need to keep track of the paths of the books that need to be updated) or each time on book close / return to filemanager (only one book handle each time but we still need to support EVT to update the PB library during the lifetime of KO within the same document)

Basically there're two different things to worry about and, even more, a lot to think about how to integrate it.  A lot of questions. Huge pain :p 

I think one of the advantages of that version of Coolreader is the tightly integration with PB system. Something that we can't do (or we don't want to do?) since we target a bunch of platforms with a single codebase.

**Comment by ploum** [comment]
Hi @pazos ,

Thanks a lot for this very complete explanation. I understand the problem. It's not a "one liner fix" so not something I should even try to attempt from scratch.

Noting that the Coolreader implementation is quite buggy. From time to time (especially when the device is under 50% on battery), CoolReader will restart a book from scratch upon reopening. Which is why I waited that Koreader was stable enough.

Anyway, I guess I will have to live without that feature for now ;-)

Thanks for your time, I appreciate it greatly.

**Comment by EastEriq** [comment]
I've added the comment to #6597, but the same is relevant here too. Regarding updating the OS percentage when books are read in KOr, recently [this](https://github.com/koreader/koreader/assets/16066531/751c5359-fe5d-40f5-993b-67728b633187) came up in the pocketbook mobiread forum, pointing to [this script on github](https://github.com/jjrrw174/PocketBook-Sync-KoReader-progress-with-PB-Library), which might be a start.
